CSparse/Double/SparseMatrix.cs

```diff
@@ -420,6 +421,143 @@ public override CompressedColumnStorage<double> Multiply(CompressedColumnStorage
             return result; // success
         }
 
+        public override CompressedColumnStorage<double> ParallelMultiply(CompressedColumnStorage<double> other)
+        {
+            // Check inputs
+            if (other == null)
+            {
+                throw new ArgumentNullException("other");
+            }
+
+            if (this.ColumnCount != other.RowCount)
+            {
+                throw new ArgumentException(Resources.MatrixDimensions);
+            }
+
+            int m = this.rowCount;
+            int n = other.ColumnCount;
+
+            if ((m > 0 && this.ColumnCount == 0) || (other.RowCount == 0 && n > 0))
+            {
+                throw new Exception(Resources.InvalidDimensions);
+            }
+
+            if (m <= 0 || n <= 0)
+            {
+                return Multiply(other);
+            }
+
+            int anz = this.NonZerosCount;
+            int bnz = other.NonZerosCount;
+
+            // Heuristics to determine whether parallel multiplication is faster
+            // Number of ops to exceed parallel overhead
+            const int min_ops = 15000;
+            // Average number of "x[i] += beta * Values[p]" per result column
+            long col_ops = (long)anz * bnz / this.ColumnCount / n;
+            if (col_ops * n < 4 * min_ops)

```

Comment:
If 4 is the number of available cores/threads, how does this affect CPUs with more/less cores. Should `Environment.ProcessorCount` be used here?

Same question for line 464 below.

Reply:
I'm still trying to find the best parameters by running benchmarks on parameter permutations. As each run takes about 45 minutes (more when restricting to 2 threads), this will take a while longer. 4 is not the number of available threads as I'm running my experiments on a system with 48 hyperthreads (2 x Intel Xeon Silver 4116).